### DISCUSSION TITLE
Add support for S3 Express One Zone

### DISCUSSION THREAD

**Discussion by cpaika** [duckdb/duckdb#9922] [Ideas]
AWS recently release [AWS S3 Express One Zone](https://aws.amazon.com/s3/storage-classes/express-one-zone/) which offers lower latency GET's/PUT's in a single AZ.

Testing S3 Express w/ DuckDB and running `read_parquet`, I'm getting this exception:
```
SELECT count(*) FROM read_parquet('s3://s3-express-test--usw2-az1--x-s3/*.parquet');
Error: IO Error: HTTP GET error on '/?encoding-type=url&list-type=2&prefix=' (HTTP 400)
```

I assume it's a slightly different API since it's AZ specific.  Ideally DuckDB would support querying/writing to S3 Express, since it opens up more use cases with the lower latency.

**Comment by cpaika** [comment]
It looks like you can get this to work by overriding the `s3_endpoint` with something like this:
`SET s3_endpoint = 's3express-usw2-az1.us-west-2.amazonaws.com';`

**Reply by szarnyasg** [reply]
This indeed does the trick. Thanks for sharing!

**Reply by hveiga** [reply]
After testing a bit, it looks like even if setting `s3_endpoint` makes it work, it does not really leverage the performance S3 Express One Zone offers. Some changes are required at the client level to get better performance, I believe in particular these two:
 - https://docs.aws.amazon.com/AmazonS3/latest/userguide/s3-express-optimizing-performance-guidelines-design-patterns.html#s3-express-optimizing-performance-session-authentication
 - https://docs.aws.amazon.com/AmazonS3/latest/userguide/s3-express-optimizing-performance-guidelines-design-patterns.html#s3-express-optimizing-performance-checksums

I am willing to give a try to contribute to get S3 Express One Zone support implemented for DuckDB. DuckDB uses [s3fs.cpp](https://github.com/duckdb/duckdb/blob/main/extension/httpfs/s3fs.cpp) to interact with AWS S3. Would it be better to extend that file or better to create a `s3expressfs.cpp` for the S3 Express specific support? I also believe some changes would be required on `duckdb_aws` extension for the session authentication. Some guidance would be greatly appreciated.

**Reply by samansmink** [reply]
hi @hveiga, thanks for your interest in contributing! Happy to provide some guidance, feel free to reach out to me directly on the DuckDB discord as well.

## Use session-based authentication
If i understand correctly, this just means to use some specially acquired temporary credentials to authenticate requests. I think it makes sense to first experiment with this first to check that the difference is actually worth it. DuckDB already supports using temporary credentials through:

```
CREATE SECRET s1 (
    TYPE S3,
    PROVIDER config,
    KEY_ID 'my_key',
    SECRET 'my_secret',
    SESSION_TOKEN 'my_session_token'
)
```

If that turns out to make a big difference I would say we want to:
- Create a new secret provider that can generate these temporary credentials from regular credentials
- Add a refresh API to DuckDB's `Secrets` for this new provider to implement

## Changing checksum 
Afaict this only applies to uploading? [this](https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html#API_GetObject_ResponseSyntax) says that the checksum is determined on upload.

I think if you need to make changes here they can just go in s3fs.cpp. Again I would first try to figure out if this actually makes a difference by uploading some files with the aws cli, confirming they use the correct checksum method and then benchmarking the difference

**Reply by hveiga** [reply]
Thanks for the reply. I'll run some tests and get back to you.
